### PR TITLE
Fix get products API for ranged products

### PR DESCRIPTION
**Pull request by selvakumar-suyambulingam**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Change Log
 
+## [2.0.0] - 2022-11-13
+### Changed
+- Incorporate all DVS API changes so far.
+
+## [1.1.0] - 2021-07-26
```
Comment (mohdhallal) [inline CHANGELOG.md:27]: Do wen need 1.1.0 given that we're going for 2.0.0?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,13 @@
 # Change Log
 
+## [2.0.0] - 2022-11-13
+### Changed
```
Comment (mohdhallal) [inline CHANGELOG.md:13]: Can we describe the changes in more details?

File: src/main/java/com/dtone/dvs/DvsApiClient.java [hunk 3]
```diff
@@ -418,20 +466,109 @@ public ApiResponse<List<Balance>> getBalances(BalanceFilter balanceFilter) throw
...
 	/**
-	 * Get statement for an account number
+	 * Get statement
 	 * 
-	 * @return statement details
+	 * @return paginated list of statement
 	 */
-	private ApiResponse<List<StatementDetail>> getStatement(StatementFilter statementFilter) throws DvsApiException {
-		return this.dvsApiClientHelper.getStatement(statementFilter);
+	private Page<ApiResponse<List<StatementDetail>>> getStatements() throws DvsApiException {
```
Comment (mohdhallal) [inline src/main/java/com/dtone/dvs/DvsApiClient.java]: Did we end up testing statement retrieval? If not, its not advisable to release it (probably remove these functions for now).

File: src/main/java/com/dtone/dvs/DvsApiClient.java [hunk 4]
```diff
@@ -418,20 +466,109 @@ public ApiResponse<List<Balance>> getBalances(BalanceFilter balanceFilter) throw
...
+
+	/**
+	 * Get campaigns by filters
+	 * 
+	 * @param campaignFilter the campaign filter object
+	 * @param recordsPerPage the number of records per page
+	 * @return list of campaigns
+	 */
+	private ApiResponse<List<Campaign>> getCampaigns(CampaignFilter campaignFilter) throws DvsApiException {
+		return this.dvsApiClientHelper.getCampaigns(campaignFilter, 0, 0);
```
Comment (mohdhallal) [inline src/main/java/com/dtone/dvs/DvsApiClient.java]: Campaigns response can still be paginated, even if a filter is used.

File: README.md [hunk 5]
```diff
@@ -267,60 +309,86 @@ This library deals with the following list of domain objects from the DVS API:
...
- 
-- Cancel Transaction:
- 
-
-    `ApiResponse<TransactionResponse> cancelTransactionResponse = dvsApiClient.cancelTransaction(transactionId);`
+        ```
+        TransactionRequest transactionRequest = new TransactionRequest();
+        transactionRequest.setExternalId("<String External Id>");
+        transactionRequest.setAutoConfirm(true);
+        transactionRequest.setProductId(20153L);
```
Comment (mohdhallal) [inline README.md:339]: `L` is a typo?

File: src/main/java/com/dtone/dvs/DvsApiClient.java [hunk 6]
```diff
@@ -172,16 +175,41 @@ public ApiResponse<List<Operator>> lookupOperators(String mobileNumber) throws D
...
+
+	/**
+	 * Get list of operators by mobile number
+	 * 
+	 * @param mobileNumber the mobile number
+	 * @return list of operators
+	 */
+	public ApiResponse<List<Operator>> lookupOperators(LookupOperatorRequest lookupOperatorRequest)
+			throws DvsApiException {
+		return this.dvsApiClientHelper.lookupOperators(lookupOperatorRequest, 0, 0);
```
Comment (mohdhallal) [inline src/main/java/com/dtone/dvs/DvsApiClient.java]: This is also a bug waiting to happen as the lookup response is paginated.

Comment (mohdhallal) [reply src/main/java/com/dtone/dvs/DvsApiClient.java]: Just noticed that you're providing another function (above) to support pagination. What is the use for `lookupOperators(LookupOperatorRequest lookupOperatorRequest)`? It seems that it enables the wrong usage of the DVS API.

File: CHANGELOG.md (hunk 1 above)
Comment (mohdhallal) [reply CHANGELOG.md:27]: nvm, 1.1.0 was released with no changelog entry, my bad

File: src/main/java/com/dtone/dvs/helper/DvsApiClientHelper.java [hunk 7]
```diff
@@ -130,6 +133,13 @@ public ApiResponse<List<Operator>> lookupOperators(String mobileNumber, int page
 				});
 	}
 
+	public ApiResponse<List<Operator>> lookupOperators(LookupOperatorRequest lookupOperatorRequest, int page,
```
Comment (mohdhallal) [inline src/main/java/com/dtone/dvs/helper/DvsApiClientHelper.java]: `page` and `recordsPerPage` are not used in this function.

**Review by mohdhallal — COMMENTED** [review]
Thanks @selvakumar-suyambulingam , i guess we should avoid opening up functions that retrieve the first page only (if my understanding is correct). The client should always manage the pagination on their side and retrieve the page/records they want.

What do you think?

File: README.md (hunk 5 above)
Comment (selvakumar-suyambulingam) [reply README.md:339]: Product Id is of type long. "L" is to say that the type is Long. We can use like Long.valueOf((20153L) as well.

File: src/main/java/com/dtone/dvs/DvsApiClient.java (hunk 3 above)
Comment (selvakumar-suyambulingam) [reply src/main/java/com/dtone/dvs/DvsApiClient.java]: It's a private function. So, we are not offering this API now.

File: src/main/java/com/dtone/dvs/DvsApiClient.java (hunk 6 above)
Comment (selvakumar-suyambulingam) [reply src/main/java/com/dtone/dvs/DvsApiClient.java]: This method is to support the new 'POST' API to fetch operators introduced by DVS API.
lookupOperatorRequest has the page parameters and mobile number.

File: src/main/java/com/dtone/dvs/DvsApiClient.java (hunk 4 above)
Comment (selvakumar-suyambulingam) [reply src/main/java/com/dtone/dvs/DvsApiClient.java]: This is applicable to all all the methods that support Filters.

We have provided inbuilt pagination only for "get all" APIs.
Clients have to use their inbuilt pagination for all other APIs.

**Comment by selvakumar-suyambulingam** [comment]
**Inbuilt pagination idea:**
We have provided inbuilt pagination only for "get all" APIs.
Clients have to use their inbuilt pagination for all other APIs. 

The pagination wrapper "Page" provides 5 functions to support pagination.
first()
next()
last()
hasNext()
hasPrevious()
The client will call first() method to get the first page records. Then, they will call next() to get the next page records.

**We can stop providing inbuilt pagination for the following reasons:**
Our other SDKs do not provide inbuilt pagination. So, we do not have to provide pagination in Java SDK.
Extra maintenance time (Bug fix, change request)
If we decide to provide pagination for APIs with filters, then, we will have to create methods with specific names which will provide too many options to the customer and customers might get confused.


Either we can simply disable the pageable methods and let the customers handle pagination.
Or else we should provide pageable methods for methods with filters as well and name all pageable methods with suffix pageable.

Thoughts?

**Comment by mohdhallal** [comment]
> 

The page wrapper is good, and we should keep it. With this wrapper in place, it functions returning the first page make much more sense.

**Pull request merged**